Core functionality of erudite micro learning web application

JayKay24 opened the pull request:
#### What does this PR do?
This PR incorporates the basic functionality of the micro-learning application which facilitates sending out a link a day to registered users about something they want to learn.

#### Screenshots
<img width="1440" alt="screen shot 2018-07-16 at 00 49 46" src="https://user-images.githubusercontent.com/20414083/42738795-78d79268-8892-11e8-8dd3-351a4a4ca9cf.png">
<img width="1440" alt="screen shot 2018-07-16 at 00 49 59" src="https://user-images.githubusercontent.com/20414083/42738796-7bcb924e-8892-11e8-8350-7bcf77c421f3.png">
<img width="1440" alt="screen shot 2018-07-16 at 00 50 12" src="https://user-images.githubusercontent.com/20414083/42738797-7e426fac-8892-11e8-9abf-d97c80da5720.png">
<img width="1440" alt="screen shot 2018-07-16 at 00 50 42" src="https://user-images.githubusercontent.com/20414083/42738801-882c9e16-8892-11e8-8d0e-f1f866e49c93.png">


#### What are the relevant pivotal tracker stories?
* [#158485020](https://www.pivotaltracker.com/story/show/158485020)
* [#158485024](https://www.pivotaltracker.com/story/show/158485024)
* [#158485050](https://www.pivotaltracker.com/story/show/158485050)
* [#158485039](https://www.pivotaltracker.com/story/show/158485039)
* [#158485036](https://www.pivotaltracker.com/story/show/158485036)

JayKay24 committed:
feat(user_management): Add signup end login endpoint

This feature creates endpoints to allow
for the registration and log in of a user
with the application.

[Delivers #158485020]

JayKay24 committed:
feat(categories): Create categories filter

This commit creates a feature to add categories
to be used to return meaningful links to pages
a user wants to learn about.

[Delivers #158485050]

JayKay24 committed:
Merge pull request #2 from JayKay24/ft-create-categories-filter-158485050

Ft create categories filter 158485050

JayKay24 committed:
feat(links): Set and receive links to categories

This commit adds the functionality for the
selected categories to receive a link
pertaining to the category.

[Delivers ##158485030]

JayKay24 committed:
Merge pull request #3 from JayKay24/ft-receieve-links-158485030

Basic functionality of erudite micro learning web application

morriskimani commented on Gemfile:
```diff
@@ -1,10 +1,22 @@
\ No newline at end of file
+gem 'sinatra-activerecord'
+gem 'sinatra-flash'
+gem 'warden', '~> 1.2', '>= 1.2.7'
```
style: Let's have gems ordered alphabetically. Make use of rubocop to flag such style issues

morriskimani commented on app/controllers/application_controller.rb:
```diff
@@ -1,9 +1,79 @@
+  configure(:development, :test) { set :session_secret, ENV['SHOTGUN_SECRET'] }
+  enable :sessions
+  register Sinatra::Flash
+  # use Rack::Session::Cookie, secret: ENV['RACK_SECRET']
```
Please remove commented out code/config.

morriskimani commented on app/controllers/category_controller.rb:
```diff
@@ -0,0 +1,33 @@
+require 'warden'
+
+class CategoryController < ApplicationController
+  get '/create_category' do
```
A better route name might be `/category/new`

morriskimani commented on app/controllers/category_controller.rb:
```diff
@@ -0,0 +1,33 @@
+    haml :'category/categories'
+  end
+
+  post '/create_category' do
```
I'd prefer this route `post '/category'`. It's more RESTful 😄 

Check out the _ROUTES OVERVIEW_ section in[ this site](https://learn.co/lessons/sinatra-restful-routes-readme)

morriskimani commented on app/controllers/link_controller.rb:
```diff
@@ -0,0 +1,63 @@
+require 'warden'
+
+class LinkController < ApplicationController
+  bing_search_api = BingService::BingWebSearchApi.new
```
👍

morriskimani commented on app/controllers/link_controller.rb:
```diff
@@ -0,0 +1,63 @@
+  bing_search_api = BingService::BingWebSearchApi.new
+
+  get '/category_links/:category_id' do
+    redirect '/login' unless current_user
```
This is becoming a bit repetitive. We have this line at the beginning of each route handler.

I suggest that you try to make use of before filters to make this less repetitive. See the filters section of [Sinatra docs](http://sinatrarb.com/intro.html)

Also, you have a method `check_authentication` in your `ApplicationController` that should do this, why are you not using it?

morriskimani commented on app/controllers/category_controller.rb:
```diff
@@ -0,0 +1,33 @@
+  post '/create_category' do
+    redirect '/login' unless current_user
+    @current_user = current_user
+    @category = @current_user.categories.create(category_name: params[:category],
```
I suggest that you use `new` here instead of `create`. `create` will actually try to save the `@category` in the database. In other words, `create` calls save internally and returns the newly created record. `new` on the other hand simply instantiates the object but doesn't save it in your DB.

Since you are calling save in the next line anyway, why not defer saving till that point.

morriskimani commented on app/controllers/link_controller.rb:
```diff
@@ -0,0 +1,63 @@
+    redirect '/login' unless current_user
+    @current_user = current_user
+
+    term = Category.where(id: params[:category_id].to_i).first
```
When searching for a record by its primary key, prefer to use `find` rather than `where`.

Advantages of using find:
- Returns a single record, not an array of records. So no need to call `.first`
- If the primary key is a numeric type, the parameter passed is coerced to an integer. So no need of calling `to_i`

[see docs](http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find)

morriskimani commented on app/controllers/link_controller.rb:
```diff
@@ -0,0 +1,63 @@
+      link: result_set[0]['displayUrl'],
+      snippet: result_set[0]['snippet']
+    )
+    category_link.save
```
Any reason why you are calling `.save` here?

morriskimani commented on app/controllers/link_controller.rb:
```diff
@@ -0,0 +1,63 @@
+
+    category = @current_user.categories.find_by(id: params[:category_id].to_i)
+
+    link_name = category.links.find_by(link_name: result_set[0]['displayUrl'])
```
Bug: This query has an error. You are finding by `link_name`, yet you are providing a `link`. `link_name` will be` nil` always because the query is incorrect.

See [how you are saving links](https://github.com/JayKay24/micro_learning_web_app/pull/4/files#diff-e50ff0c44e418c0730178bd65a9498e0R46) if you are not convinced

morriskimani commented on app/controllers/link_controller.rb:
```diff
@@ -0,0 +1,63 @@
+
+    until link_name.nil?
+      result_set.shuffle!
+      link_name = category.links.find_by(link_name: result_set[0]['displayUrl'])
```
Bug: **Possible Infinite Loop.** The reason you cannot see the bug can be attributed to the error I've highlighted above.

Imagine a scenario where `result_set[0]['displayUrl']` has already been visited, and thus saved in your database. In such a situation, you will have an infinite loop. Because you are providing the exact same search result (the one indexed at **0**). I'd have expected you to increment that index on every iteration.

Please review this logic.

morriskimani commented on app/controllers/link_controller.rb:
```diff
@@ -0,0 +1,63 @@
+      snippet: result_set[0]['snippet']
+    )
+    category_link.save
+    @category_links = [result_set[0]]
```
This follows the assumption that calling `shuffle!` on an array will always result in a change of what is stored in the first position. Is that always the case?

morriskimani commented on app/controllers/user_controller.rb:
```diff
@@ -1,13 +1,51 @@
+  end
+
+  post '/signup' do
+    if /^(?=.*[a-zA-Z])(?=.*[0-9])(?=.*[\W]).{8,}$/.match?(params[:password])
```
This logic does not belong in the controller. I believe you can have it as a validation in your User model

morriskimani commented on app/controllers/user_controller.rb:
```diff
@@ -1,13 +1,51 @@
+
+  post '/signup' do
+    if /^(?=.*[a-zA-Z])(?=.*[0-9])(?=.*[\W]).{8,}$/.match?(params[:password])
+      @user = User.new(first_name: params[:first_name],
```
I'd suggest that you have a private method that extracts user params. That way, we reduce the work being done by this route handler. If you do that, you can simply have:
```ruby
@user = User.new(user_params)
```

morriskimani reviewed (changes requested):
Good start @JayKay24, I left a few comments though. Please have a look.

Note: some of the issues I have raised occur more than once. Please fix all instances of such issues.

morriskimani commented on app/controllers/link_controller.rb:
```diff
@@ -0,0 +1,63 @@
+    redirect '/login' unless current_user
+    @current_user = current_user
+
+    term = @current_user.categories.find_by(id: params[:category_id].to_i)
```
When searching using `primary key`, prefer to use `find` rather than `find_by` or `where`

```ruby
 term = @current_user.categories.find(params[:category_id])
```

JayKay24 committed:
chore(core-functionality): Test the core functionality of the
application

This commit adds all the basic tests for the core functionality of
the application.

[Delivers #159071954]

JayKay24 committed:
Merge pull request #5 from JayKay24/ch-tests-for-core-functionality-159071954

Test all routes and core functionality of the application
